src/app/data-planes/data/index.ts

```diff
@@ -8,3 +12,40 @@ export function getLastUpdateTime(subscriptions: DiscoverySubscription[]): strin
   const lastSubscription = subscriptions[subscriptions.length - 1]
   return lastSubscription.status.lastUpdateTime
 }
+
+export function getStatusAndReason(dataplaneOverview: DataplaneOverview): { status: StatusKeyword, reason: string[] } {
+  const inbounds = dataplaneOverview.dataplane.networking.inbound ?? []
+  const errors = inbounds
+    .filter((inbound) => inbound.health && !inbound.health.ready)
+    .map((inbound) => `Inbound on port ${inbound.port} is not ready (kuma.io/service: ${inbound.tags['kuma.io/service']})`)
+
+  let status: StatusKeyword
+  switch (true) {
+    case inbounds.length === 0:
+      status = 'online'
+      break
+    // If errors and inbounds are equal (even if they are both 0) then we are offline
+    case errors.length === inbounds.length:
+      status = 'offline'
+      break
+    // If there are any errors then we are degraded
+    case errors.length > 0:
+      status = 'partially_degraded'
+      break
+    default: {
+      const subscriptions = dataplaneOverview.dataplaneInsight?.subscriptions ?? []
+
+      if (subscriptions.length === 0) {
+        status = 'offline'
+      } else {
+        const lastSubscription = subscriptions[subscriptions.length - 1]
+        status = lastSubscription.connectTime?.length && !lastSubscription.disconnectTime ? 'online' : 'offline'
+      }
+    }
```

Comment:
> Could you move `zones/data#getStatus` over to `subscriptions/data#getStatus` and then reuse that here?

I’ve done that now and added tests for it.

> Actually could we rename getStatus to getState while we are doing that?

I kept the status naming, I think it makes more sense and we virtually always refer to this information as status already.

Reply:
It’s not so much about the effort this would take, but the name being worse in my opinion. It shouldn’t be called state because what we’re looking to get out of this function is the status of a resource. State is too loose of a term for this.

The fact that there is a `status` field on KDS subscriptions is unfortunate, but I’d rather call the thing that’s a resource’s status by that name than the thing that would more accurately be called `stat` or `stats` (as is the case for Discovery subscriptions).

Reply:
Yeah, I think `getIsConnected` is much better. It’s also nice that it no longer returns `'online' | 'offline'` and instead the resource-specific functions own those keywords completely.